src/utils/series.ts

```diff
@@ -1,85 +1,114 @@
 import type { Playlist, PlaylistItem } from '#types/playlist';
-import type { EpisodeWithSeason, Season, Series } from '#types/series';
+import type { EpisodesWithPagination, Series } from '#types/series';
 
-export const getFiltersFromSeries = (playlist: Playlist): string[] =>
-  playlist.playlist.reduce(
-    (filters: string[], item) => (item.seasonNumber && filters.includes(item.seasonNumber) ? filters : filters.concat(item.seasonNumber || '')),
-    [],
-  );
+/**
+ * Get an array of options for a season filter
+ */
+export const getFiltersFromSeries = (playlist: Playlist, series: Series | undefined): string[] => {
+  const isNewFlow = !!series;
+
+  // For the new series flow we already have episodes sorted correctly on the back-end side
+  if (isNewFlow) {
+    return 'seasons' in series ? series.seasons.map((season) => String(season.season_number)) : [];
+  }
+
+  // Old series doesn't have sorting supported and just aggregates all episodes in one playlist
+  // So we need to sort the playlist manually based on the selected filter (season).
+  return playlist.playlist
+    .reduce((filters: string[], item) => (item.seasonNumber && filters.includes(item.seasonNumber) ? filters : filters.concat(item.seasonNumber || '')), [])
+    .slice()
+    .sort();
+};
+
+/**
+ * Get a playlist with episodes based on the selected filter
+ */
+export const filterSeries = (playlist: Playlist, episodes: EpisodesWithPagination[] | undefined, filter: string): Playlist => {
+  const isNewFlow = !!episodes?.length;
+
+  if (isNewFlow) {
+    // Get a flattened list of episodes and return it as part of the playlist
+    return { ...playlist, playlist: episodes.flatMap((e) => e.episodes) };
+  }
 
-export const filterSeries = (playlist: Playlist, filter: string) => {
   if (!filter) return playlist;
 
   return {
     ...playlist,
+    // Filter episodes manually for the old flow where our playlists includes all episodes with all seasons
     playlist: playlist.playlist.filter(({ seasonNumber }) => seasonNumber === filter),
   };
 };
 
-export const getSeriesEpisodes = (series: Series) => {
-  if ('seasons' in series) {
-    return series.seasons.flatMap((season: Season) => season.episodes);
+export const getNextEpisode = (data: EpisodesWithPagination[], currentEpisode: PlaylistItem, hasMoreEpisodes: boolean, fetchNextEpisodes: () => void) => {
+  const episodes = data.flatMap((el) => el.episodes);
+  const episodeIndex = episodes.findIndex((el) => el.mediaid === currentEpisode.mediaid);
+
+  // If there are more episodes in the selected season / episodes list
+  if (episodeIndex !== episodes.length - 1) {
+    return episodes[episodeIndex + 1];
+    // Episode is the last one in the data we have, but we have more to fetch, let's do it
+  } else if (hasMoreEpisodes) {
+    fetchNextEpisodes();
```

Comment:
What happens if the paging doesn't complete before the previous episode finishes?

Reply:
We can move this function to the 'onComplete' handler to wait for the request to resolve. In case we don't have a next episode, then we can just close the player and show the page.

Reply:
Made a separate method to fetch it.